llrt_modules/src/modules/zlib.rs

```diff
@@ -0,0 +1,293 @@
+    Compression,
+};
+use llrt_utils::{ctx::CtxExtension, module::export_default, object::ObjectExt, result::ResultExt};
+use rquickjs::function::Func;
+use rquickjs::{
+    module::{Declarations, Exports, ModuleDef},
+    prelude::{Opt, Rest},
+    Ctx, Error, Exception, FromJs, Function, IntoJs, Null, Object, Result, Value,
+};
+use zstd::{
+    stream::read::{Decoder as ZstdDecoder, Encoder as ZstdEncoder},
+    zstd_safe::CompressionLevel,
+};
+
+use crate::{utils::array_buffer::ArrayBufferView, ModuleInfo};
+
+use super::buffer::Buffer;
+
+macro_rules! define_sync_function {
+    ($fn_name:ident, $converter:expr, $command:expr) => {
+        pub fn $fn_name<'js>(
+            ctx: Ctx<'js>,
+            value: Value<'js>,
+            options: Opt<Object<'js>>,
+        ) -> Result<Value<'js>> {
+            $converter(ctx.clone(), value, options, $command)
+        }
+    };
+}
+
+macro_rules! define_cb_function {
+    ($fn_name:ident, $converter:expr, $command:expr) => {
+        pub fn $fn_name<'js>(
+            ctx: Ctx<'js>,
+            value: Value<'js>,
+            args: Rest<Object<'js>>,
+        ) -> Result<()> {
+            let mut args_iter = args.0.into_iter().rev();
+            let cb: Function = args_iter
+                .next()
+                .and_then(|v| <rquickjs::Value<'_> as Clone>::clone(&v).into_function())
+                .or_throw_msg(&ctx, "Callback required")?;
+            let options = match args_iter.next() {
+                Some(v) => Opt(Some(v)),
+                None => Opt(None),
+            };
+
+            ctx.clone().spawn_exit(async move {
+                match $converter(ctx.clone(), value, options, $command) {
+                    Ok(obj) => {
+                        () = cb.call((Null.into_js(&ctx), obj))?;
+                        Ok::<_, Error>(())
+                    },
+                    Err(err) => {
+                        () = cb.call((Exception::from_message(ctx, &err.to_string()),))?;
+                        Ok(())
+                    },
+                }
+            })?;
+            Ok(())
+        }
+    };
+}
+
+enum ZlibCommand {
+    Deflate,
+    DeflateRaw,
+    Gzip,
+    Inflate,
+    InflateRaw,
+    Gunzip,
+}
+
+fn zlib_converter<'js>(
+    ctx: Ctx<'js>,
+    value: Value<'js>,
+    options: Opt<Object<'js>>,
+    command: ZlibCommand,
+) -> Result<Value<'js>> {
+    let src = if value.is_string() {
+        let string = value.as_string().unwrap().to_string()?;
+        string.as_bytes().to_vec()
+    } else {
+        let buffer = ArrayBufferView::from_js(&ctx, value)?;
+        buffer.as_bytes().unwrap().to_vec()
+    };
+
+    let mut level = Compression::default();
+    if let Some(options) = options.0 {
+        if let Some(opt) = options.get_optional("level")? {
+            level = Compression::new(opt);
+        }
+    }
+
+    let mut dst: Vec<u8> = Vec::with_capacity(src.len());
```

Comment:
Do we need this? Doesn't the encoders reserve based on the source size? Please verify! if not, there are algorithms to calculate approximate size for most compression formats.

Reply:
I use the same library in our codebase, it is a stream based API so it only allocates a small buffer internally but you have to provide the aggregate buffer.